Add support for adding notes to Runs

aadamson opened the pull request:
Follow up for #303 

Remaining items:
- [x] Figure out what to do about error handling on `setTagApi` failure

aadamson committed:
WIP

aadamson commented:
Link to referenced comment: https://github.com/mlflow/mlflow/issues/303#issuecomment-415304458

@mateiz I went ahead and implemented a rough version of what was discussed in #303 last week before the `set_tag` API landed and only just now saw your comment (whoops). If you think it would simplify things/be consistent with your vision of how tags should be used, I'll go ahead and fold the `Note` entity into tags.

aadamson commented:
In reference to https://github.com/mlflow/mlflow/issues/303#issuecomment-415915466

@gioa Thanks for the feedback! Yeah, I think simplifying it to just one note makes sense. One use-case I was hoping to capture via the notes thing was adding a nice way to display rich, Markdown-formatted programmatic reports (like tear sheets with collated learning curves for different metrics) which there in theory could be many of, but that would probably be better addressed with some other feature or even just adding support for rendering Markdown in the artifact viewer.

aadamson commented:
@mateiz I'm going to go ahead and reimplement this on top of tags. My plan is to follow along with @gioa's suggestion and have just a single note, so the tag would have the key "mlflow.note". Tags that start with "mlflow." would be filtered out of the current tags view on the run page since they're assumed to have some other metadata meaning that is consumed elsewhere in the UI. Does this make sense?

aadamson commented:
@gioa I sat down with the user who was asking for this, and he was amenable to having just a single note. I think your mockups are a lot cleaner than what I had, so I'm going to go ahead and follow your lead here

mateiz commented:
OK, that sounds great. We can eventually add an API for appending to the note too if that helps. And yes, we do intend to filter tags with `mlflow.` out of the view and just use them for features we display in other ways.

aadamson commented:
I reimplemented it just using `set_tag`. I think it would make sense to punt on adding any Python API support to another PR. 

One question I still have is: is there a canonical way to handle HTTP errors in the UI? If the call to `setTagApi` fails, is there an existing UI component I can use?

mateiz commented:
I tried this out and it looks pretty nice, but I noticed a couple of issues:
1) The edit box always opens in the "Code" view, not the "Preview" view that shows the rendered Markdown. It would be better if Notes actually showed the Markdown by default, and you could press an Edit button to edit it and then Save to save. This would also make the note smaller in the default case when there is no note, so people can see the rest of the page (metrics and params).
2) There's a problem (likely in the file store already, not part if you PR) where if the note has multiple lines of text, the server fails with a "tag recorded more than once" error. It also seems to have problems with special characters in the string sometimes, which make it think that it's not Unicode. @andrewmchen / @mparkhe can you guys look into this?

aadamson commented:
@mateiz I'm trying to make sure I understand your intention from your comment - what do you think should be shown if there is no note? Should the section just be collapsed (like Tags section with no tags) and then when you expand it, there's an empty preview and a button to edit?

aadamson commented:
No note uploaded:
![image](https://user-images.githubusercontent.com/3607110/45116300-4d5fe480-b120-11e8-9769-89bb90041bfb.png)

No note uploaded, but text changed:
![image](https://user-images.githubusercontent.com/3607110/45116317-59e43d00-b120-11e8-8afd-2c99b81cc7d5.png)

Note uploaded, but text unchanged:
![image](https://user-images.githubusercontent.com/3607110/45116363-741e1b00-b120-11e8-81cd-610fce7c2729.png)

It could also make sense to collapse the whole "Notes" section by default if no tag is uploaded, like we do with the "tags" section.

gioa commented:
@aadamson Thank you for the mockups, it works great! It probably looks better having the "Code (or Edit)|Preview" as a tab inside the editor, like your previous prototype in another PR. 

![adding_v1](https://user-images.githubusercontent.com/1960721/45120086-04089880-b112-11e8-8f0a-528382827b88.png)

![adding_preview_v1](https://user-images.githubusercontent.com/1960721/45120085-04089880-b112-11e8-9280-b28c3941a584.png)

I also put together a clickable prototype here to demostrate the flow. Let me know what you think!
https://invis.io/87NWFUUHBZJ#/318321280_empty_V1

aadamson committed:
Hook up to reducer

aadamson commented:
@gioa What should "Cancel" button do if there is already an existing note? Discard any edits and go back to preview mode and keep the notes view exposed?

gioa commented:
Thank you for the comments! Let me list all the mockups here to explain more about the flow in the InVision prototype.

- No notes at the beginning 

![empty_v1](https://user-images.githubusercontent.com/1960721/45174355-7e432680-b1bf-11e8-8d03-c7661ca7c6e9.png)

- User clicks on an edit icon to trigger the editor, default on code tab.

![adding_v1](https://user-images.githubusercontent.com/1960721/45174402-a3d03000-b1bf-11e8-8bc7-ea1baab48c6c.png)

- User edits the notes and clicks on the editor's preview tab to make sure the changes seem right.

![adding_preview_v1](https://user-images.githubusercontent.com/1960721/45174463-ce21ed80-b1bf-11e8-8362-5a7adddc3bec.png)

- User clicks "Save" to upload the changes or "Cancel" to discard the changes. Both go back to a read-only mode as below (no editor, but has a editor button to trigger the editor).   

![done_v1](https://user-images.githubusercontent.com/1960721/45174621-2f49c100-b1c0-11e8-8834-a0d74b2cc3ba.png)

Hope this helps. What do you think?

aadamson commented:
Oh yep, I misunderstood you, I thought your mockups only applied to the case where there was no note saved. Thanks for the help @gioa, I'll fix it up and circle back!

aadamson committed:
Have preview shown by default; click to open editor

aadamson commented:
I tried to address your comments @gioa, let me know what else I need to change!

@andrewmchen I'm getting the following stack trace once I've saved a multi-line tag:
```bash
[2018-09-06 14:32:59,917] ERROR in app: Exception on /ajax-api/2.0/preview/mlflow/runs/search [POST]
Traceback (most recent call last):
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/_compat.py", line 35, in reraise
    raise value
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/alex/miniconda3/envs/mlflow_dev/lib/python3.6/site-packages/Flask-1.0.2-py3.6.egg/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/alex/twosigma/mlflow/mlflow/server/handlers.py", line 208, in _search_runs
    request_message.anded_expressions)
  File "/Users/alex/twosigma/mlflow/mlflow/store/file_store.py", line 380, in search_runs
    return [self.get_run(run_uuid) for run_uuid in run_uuids]
  File "/Users/alex/twosigma/mlflow/mlflow/store/file_store.py", line 380, in <listcomp>
    return [self.get_run(run_uuid) for run_uuid in run_uuids]
  File "/Users/alex/twosigma/mlflow/mlflow/store/file_store.py", line 250, in get_run
    tags = self.get_all_tags(run_uuid)
  File "/Users/alex/twosigma/mlflow/mlflow/store/file_store.py", line 361, in get_all_tags
    tags.append(self._get_tag_from_file(parent_path, tag_file))
  File "/Users/alex/twosigma/mlflow/mlflow/store/file_store.py", line 339, in _get_tag_from_file
    % tag_name)
Exception: Unexpected data for tag 'mlflow.note.content'. Tag recorded more than once
```
The tags file it is complaining about looks like
```bash
asdfasdf asdf
sdfasdf
```
(file is  mlruns/0/38b05370ce3e4e3c8485f3081c465614/tags/mlflow.note.content). What do you think we can do to address this? What are the semantics of multi-line tags supposed to be? Should I just be encoding the tag before uploading it so it is a single line?

aadamson committed:
Fix lint issues

gioa commented:
@aadamson Thank you for the quick turnaround! Looks great! 
A few nits:
1. Preview tab does not work for some formats. For example, ~~test~~ does not have strikethrough until the changes are saved. 
2. Playing with the editor I found it hard to notice the switch between Code and Preview tab. Can we add a "border: 2px solid #0092E1" to the class ".mde-text" to indicate its editable state?
3. Do we need a min-height of 200px for note-view-preview-content? It seems to me we can get rid of that.  
4. Can we follow the styles for the Save button? Both background and border color is #0193e1 by default and when hover it's #017fcb.

Thank you!

aadamson committed:
Address style issues; set MD flavor to github

aadamson committed:
Support multiline tags

aadamson commented:
Thanks! Tried to address your comments. The color you provided for the "save" button looks a lot darker than it does in your mockups, but of course I can't see the styles in your mocks. Let me know what else I can do.

gioa commented:
Thank you for addressing the comments! Appreciate your attention to details!

- about the button style, it uses the bootstrap style somehow. 
- could you also remove the extra 10px padding-bottom for the <p> under "note-view-preview-content" so the spacing looks equal? It comes from bootstrap's type.less. 
- I saw you adjusted the spacing between Save and Cancel. Looks great! Could you also change div.btn-toolbar padding-top from 4px to 16px?

gioa commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -264,6 +330,17 @@ class RunView extends Component {
+  // eslint-disable-next-line no-unused-vars
+  handleSubmittedNote(content, err) {
+    if (err) {
+      // TODO(adamson) figure out what to display on an error
```
Show the error message above the buttons? See below. Also would be great not leaving extra spacing for the error messages but pushing other content down when there is an error in the page. 

![adding_preview_error](https://user-images.githubusercontent.com/1960721/45193495-1ca4ab80-b203-11e8-8e36-c286fb0ee173.png)

aadamson replied:
I added an alert above the buttons when an error occurs.

aadamson committed:
Add alert box on error

aadamson committed:
Address python lint issues

aadamson committed:
Remove TODO

aadamson commented:
Hey guys, I think this is ready for final review. @andrewmchen, can I get your eyes on my changes to the file store? I wasn't sure if the lack of support for multi-line tags was intentional, but it didn't seem to be.

andrewmchen commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -98,8 +125,45 @@ class RunView extends Component {
   }
 
+  renderNoteSection(noteInfo) {
+    if (this.state.showNotes) {
```
We should make sure that the expander can not be clicked when the notes editor is being shown. In the current implementation, if we close the expander, the state in the `NoteEditorView` will be lost since we are unmounting that component.

Likewise, we should make sure that any route changes are protected with a Prompt. I believe you can do this with this component. https://reacttraining.com/react-router/core/api/Prompt

aadamson replied:
I tried making it so that RunView would store the state of the editor whenever it was going to get unmounted and then restore it on next render, but that got really hairy without using refs.

Should the `onclick` on the expander just be disabled and the caret hidden when the editor is shown? That feels a bit counterintuitive, but I don't have a better, simple idea.

andrewmchen replied:
Yeah, let's just disable the onClick on the expander.

aadamson replied:
Done

andrewmchen reviewed (changes requested):
Looks good in general. The comments are mostly about how we address toggling the "notes" header when the editor is open. (Currently we may lose some state.)

aadamson commented on mlflow/store/file_store.py:
```diff
@@ -332,12 +332,7 @@ def _get_param_from_file(parent_path, param_name):
-            raise Exception("Unexpected data for tag '%s'. Tag recorded more than once"
-                            % tag_name)
-        return RunTag(tag_name, str(tag_data[0].strip()))
+        return RunTag(tag_name, str(tag_data.strip()))
```
Yeah I agree, there's nothing on my end that necessitates stripping, and it would make more sense if tag data weren't stripped in the context of this PR

aadamson replied:
I went ahead and removed the call to `trim`

aadamson replied:
I also had to modify `set_tag` in `FileStore` to not add a trailing newline (since otherwise the tests failed because there was an unexpected trailing newline that was not stripped out). I'm happy to revert these I don't care that much.

aadamson replied:
Also worth noting that these changes are not backwards-compatible w.r.t. how existing notes would be displayed. I'm not sure what your policy is about that during alpha.

aadamson commented on mlflow/server/js/src/utils/NoteUtils.js:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  static fromRunTags = (tags) => {
+    const obj = Object.values(tags).map((t) =>
```
Yep, definitely

aadamson committed:
Address javascript comments

aadamson committed:
Dont strip tag values

aadamson committed:
Remove unused NoteView.css

aadamson committed:
Factor out markdown converter; add CSS for preview

aadamson committed:
Remove unused react-markdown; explain CSS in NoteView.css

aadamson committed:
Address comments

aadamson commented:
@andrewmchen I tried to address all of your comments. I also went ahead and addressed an issue that I just noticed, which is that the markdown generator and the CSS for the markdown-generated HTML were way off between the preview in the editor and the actual note that gets rendered. I addressed by making the MD->HTML converters match and by adapting the CSS from the component that renders the preview (react-mde) to the component that renders when you're not in the editor and pinning react-mde.

aadamson committed:
Lint

aadamson committed:
Do not add trailing newline to tag data

aadamson commented:
I think the failure is a flake - can you rerun the build @andrewmchen? The test passes locally and there's no clear connection (to me) to my changes

andrewmchen commented on mlflow/server/js/src/components/RunView.js:
```diff
@@ -214,6 +264,21 @@ class RunView extends Component {
         <div className="RunView-info">
+          <h2 className="table-name">
+            <span
+              onClick={this.state.showNotesEditor ? false : () => this.onClickExpander(NOTES_KEY)}
```
Some react warnings in the JS console.

```If you used to conditionally omit it with onClick={condition && value}, pass onClick={condition ? value : undefined} instead.```

aadamson committed:
Fix console warnings

andrewmchen reviewed (changes requested):
Code looks mostly good! My last major concern is that this feature is vulnerable to stored xss. Try making this your note :)

```
<img src="/static-files/static/media/home-logo.7a9dd62d.png" onload="alert('test');" />
```

aadamson commented on mlflow/server/js/src/utils/NoteUtils.js:
```diff
@@ -0,0 +1,17 @@
+import { MLFLOW_INTERNAL_PREFIX } from "./TagUtils";
+
+export const NOTE_TAG_PREFIX = MLFLOW_INTERNAL_PREFIX + 'note.';
```
Ambition and plans to store other stuff like `MLFLOW_INTERNAL_PREFIX + 'note.last_edited_at'` haha. I'll go ahead and simplify it

aadamson commented:
@andrewmchen do you have an opinion on what to do about the XSS vulnerability? Should I be doing server- or client-side stripping of any and all HTML tags or things that could be converted into valid HTML from the user's input? I'll try to find some prior art (e.g. how Github pulls it off)

aadamson committed:
Address comments

andrewmchen commented:
@aadamson, ah, I forgot to put that in. I believe the state of the art is to use https://www.npmjs.com/package/sanitize-html.

aadamson committed:
Sanitize converted-from-markdown HTMLg

aadamson committed:
Fix lint

aadamson commented:
It looks like the build just timed out. I'm not sure what that's about.

andrewmchen commented:
Yeah looks weird. You can always retrigger the build by pushing an empty commit (that way you don't have to wait on me)! I'll retrigger it this time for ya.

andrewmchen commented:
This looks good to me from a code perspective! @gioa, do you want to final sign off on the design?

gioa commented:
LGTM! Thank you for the hard work there! @aadamson

andrewmchen commented:
Thanks for the contribution! @aadamson. Looking forward to using it myself.

aadamson commented:
Thanks @andrewmchen and @gioa for all the helpful feedback! Looking forward to working more with you both soon!
